### ISSUE TITLE
`workflow.status` doesn't work with ternary operator in arguments

### ISSUE THREAD

**Issue by nikita-akuity** [argoproj/argo-workflows#10393]
### Pre-requisites

- [X] I have double-checked my configuration
- [X] I can confirm the issues exists when I tested with `:latest`
- [ ] I'd like to contribute the fix myself (see [contributing guide](https://github.com/argoproj/argo-workflows/blob/master/docs/CONTRIBUTING.md))

### What happened/what you expected to happen?

A following ternary operator in the exit handler`'{{= workflow.status == "Succeeded" ? "SUCCESSFUL" : "FAILED" }}'` results in "FAILED" even when workflow.status is "Succeeded".
Other workflow parameters seem to be processed correctly: I tested it with `workflow.parameters.foo` instead of status and it behaved as expected.

### Version

v3.4.4

### Paste a small workflow that reproduces the issue. We must be able to run the workflow; don't enter a workflows that uses private images.

```YAML
spec:
  entrypoint: run-test
  templates:
    - name: run-test
      container:
        name: runner
        image: alpine
        command: ['sh','-c']
        args:
          - exit 0
    - name: cowsay
      inputs:
        parameters:
          - name: ternary
          - name: status
      container:
        image: argoproj/argosay:v1
        command: ['sh','-c']
        args:
          - cowsay "{{inputs.parameters.ternary}} {{inputs.parameters.status}}"
  hooks:
    exit:
      template: cowsay
      arguments:
        parameters:
          - name: ternary
            value: '{{= workflow.status == "Succeeded" ? "SUCCESSFUL" : "FAILED" }}'
          - name: status
            value: '{{= workflow.status }}'
```
Output logs:
```
lovely-rhino: time="2023-01-24T07:29:32.402Z" level=info msg="sub-process exited" argo=true error="<nil>"
lovely-rhino-cowsay-1502533312:  __________________
lovely-rhino-cowsay-1502533312: < FAILED Succeeded >
lovely-rhino-cowsay-1502533312:  ------------------
lovely-rhino-cowsay-1502533312:         \   ^__^
lovely-rhino-cowsay-1502533312:          \  (oo)\_______
lovely-rhino-cowsay-1502533312:             (__)\       )\/\
lovely-rhino-cowsay-1502533312:                 ||----w |
lovely-rhino-cowsay-1502533312:                 ||     ||
lovely-rhino-cowsay-1502533312: time="2023-01-24T07:29:42.139Z" level=info msg="sub-process exited" argo=true error="<nil>"
```

### Logs from the workflow controller

```text
kubectl logs -n argo deploy/workflow-controller | grep lovely-rhino

time="2023-01-24T07:29:28.829Z" level=info msg="Processing workflow" namespace=test workflow=lovely-rhino
time="2023-01-24T07:29:28.851Z" level=info msg="Updated phase  -> Running" namespace=test workflow=lovely-rhino
time="2023-01-24T07:29:28.851Z" level=info msg="Pod node lovely-rhino initialized Pending" namespace=test workflow=lovely-rhino
time="2023-01-24T07:29:28.897Z" level=info msg="Created pod: lovely-rhino (lovely-rhino)" namespace=test workflow=lovely-rhino
time="2023-01-24T07:29:28.897Z" level=info msg="TaskSet Reconciliation" namespace=test workflow=lovely-rhino
time="2023-01-24T07:29:28.897Z" level=info msg=reconcileAgentPod namespace=test workflow=lovely-rhino
time="2023-01-24T07:29:28.910Z" level=info msg="Workflow update successful" namespace=test phase=Running resourceVersion=7190628 workflow=lovely-rhino
time="2023-01-24T07:29:38.899Z" level=info msg="Processing workflow" namespace=test workflow=lovely-rhino
time="2023-01-24T07:29:38.899Z" level=info msg="Task-result reconciliation" namespace=test numObjs=0 workflow=lovely-rhino
time="2023-01-24T07:29:38.899Z" level=info msg="node changed" namespace=test new.message= new.phase=Succeeded new.progress=0/1 nodeID=lovely-rhino old.message= old.phase=Pending old.progress=0/1 workflow=lovely-rhino
time="2023-01-24T07:29:38.900Z" level=info msg="TaskSet Reconciliation" namespace=test workflow=lovely-rhino
time="2023-01-24T07:29:38.900Z" level=info msg=reconcileAgentPod namespace=test workflow=lovely-rhino
time="2023-01-24T07:29:38.900Z" level=info msg="Running OnExit handler: " namespace=test workflow=lovely-rhino
time="2023-01-24T07:29:38.900Z" level=info msg="Pod node lovely-rhino-1502533312 initialized Pending" namespace=test workflow=lovely-rhino
time="2023-01-24T07:29:38.927Z" level=info msg="Created pod: lovely-rhino.onExit (lovely-rhino-cowsay-1502533312)" namespace=test workflow=lovely-rhino
time="2023-01-24T07:29:38.938Z" level=info msg="Workflow update successful" namespace=test phase=Running resourceVersion=7190690 workflow=lovely-rhino
time="2023-01-24T07:29:48.928Z" level=info msg="Processing workflow" namespace=test workflow=lovely-rhino
time="2023-01-24T07:29:48.929Z" level=info msg="Task-result reconciliation" namespace=test numObjs=0 workflow=lovely-rhino
time="2023-01-24T07:29:48.929Z" level=info msg="node changed" namespace=test new.message= new.phase=Succeeded new.progress=0/1 nodeID=lovely-rhino-1502533312 old.message= old.phase=Pending old.progress=0/1 workflow=lovely-rhino
time="2023-01-24T07:29:48.929Z" level=info msg="node unchanged" namespace=test nodeID=lovely-rhino workflow=lovely-rhino
time="2023-01-24T07:29:48.929Z" level=info msg="TaskSet Reconciliation" namespace=test workflow=lovely-rhino
time="2023-01-24T07:29:48.929Z" level=info msg=reconcileAgentPod namespace=test workflow=lovely-rhino
time="2023-01-24T07:29:48.929Z" level=info msg="Running OnExit handler: " namespace=test workflow=lovely-rhino
time="2023-01-24T07:29:48.929Z" level=info msg="Updated phase Running -> Succeeded" namespace=test workflow=lovely-rhino
time="2023-01-24T07:29:48.929Z" level=info msg="Marking workflow completed" namespace=test workflow=lovely-rhino
time="2023-01-24T07:29:48.929Z" level=info msg="Checking daemoned children of " namespace=test workflow=lovely-rhino
time="2023-01-24T07:29:48.935Z" level=info msg="cleaning up pod" action=deletePod key=test/lovely-rhino-1340600742-agent/deletePod
time="2023-01-24T07:29:48.957Z" level=info msg="Workflow update successful" namespace=test phase=Succeeded resourceVersion=7190754 workflow=lovely-rhino
time="2023-01-24T07:29:48.958Z" level=info msg="Queueing Succeeded workflow test/lovely-rhino for delete in 10m0s due to TTL"
time="2023-01-24T07:29:53.963Z" level=info msg="cleaning up pod" action=deletePod key=test/lovely-rhino/deletePod
time="2023-01-24T07:29:53.963Z" level=info msg="cleaning up pod" action=deletePod key=test/lovely-rhino-cowsay-1502533312/deletePod
```

### Logs from in your workflow's wait container

```text
lovely-rhino-f9r5l: time="2023-01-24T07:39:44.473Z" level=info msg="Starting Workflow Executor" version=v3.4.4
lovely-rhino-f9r5l: time="2023-01-24T07:39:44.477Z" level=info msg="Using executor retry strategy" Duration=1s Factor=1.6 Jitter=0.5 Steps=5
lovely-rhino-f9r5l: time="2023-01-24T07:39:44.477Z" level=info msg="Executor initialized" deadline="2023-01-24 09:39:42 +0000 UTC" includeScriptOutput=false namespace=test podName=lovely-rhino-f9r5l template="{\"name\":\"run-test\",\"inputs\":{},\"outputs\":{},\"metadata\":{},\"container\":{\"name\":\"runner\",\"image\":\"alpine\",\"command\":[\"sh\",\"-c\"],\"args\":[\"exit 0\"],\"resources\":{}}}" version="&Version{Version:v3.4.4,BuildDate:2022-11-29T16:49:53Z,GitCommit:3b2626ff900aff2424c086a51af5929fb0b2d7e5,GitTag:v3.4.4,GitTreeState:clean,GoVersion:go1.18.8,Compiler:gc,Platform:linux/amd64,}"
lovely-rhino-f9r5l: time="2023-01-24T07:39:44.478Z" level=info msg="Starting deadline monitor"
lovely-rhino-f9r5l: time="2023-01-24T07:39:46.479Z" level=info msg="Main container completed" error="<nil>"
lovely-rhino-f9r5l: time="2023-01-24T07:39:46.479Z" level=info msg="No Script output reference in workflow. Capturing script output ignored"
lovely-rhino-f9r5l: time="2023-01-24T07:39:46.479Z" level=info msg="No output parameters"
lovely-rhino-f9r5l: time="2023-01-24T07:39:46.479Z" level=info msg="No output artifacts"
lovely-rhino-f9r5l: time="2023-01-24T07:39:46.479Z" level=info msg="Deadline monitor stopped"
lovely-rhino-f9r5l: time="2023-01-24T07:39:46.479Z" level=info msg="stopping progress monitor (context done)" error="context canceled"
lovely-rhino-f9r5l: time="2023-01-24T07:39:46.479Z" level=info msg="Alloc=6098 TotalAlloc=12048 Sys=18386 NumGC=4 Goroutines=4"lovely-rhino-f9r5l-cowsay-100404247: time="2023-01-24T07:39:54.519Z" level=info msg="Starting Workflow Executor" version=v3.4.4
lovely-rhino-f9r5l-cowsay-100404247: time="2023-01-24T07:39:54.522Z" level=info msg="Using executor retry strategy" Duration=1s Factor=1.6 Jitter=0.5 Steps=5
lovely-rhino-f9r5l-cowsay-100404247: time="2023-01-24T07:39:54.522Z" level=info msg="Executor initialized" deadline="2023-01-24 09:39:42 +0000 UTC" includeScriptOutput=false namespace=test podName=lovely-rhino-f9r5l-cowsay-100404247 template="{\"name\":\"cowsay\",\"inputs\":{\"parameters\":[{\"name\":\"ternary\",\"value\":\"FAILED\"},{\"name\":\"status\",\"value\":\"Succeeded\"}]},\"outputs\":{},\"metadata\":{},\"container\":{\"name\":\"\",\"image\":\"argoproj/argosay:v1\",\"command\":[\"sh\",\"-c\"],\"args\":[\"cowsay \\\"FAILED Succeeded\\\"\"],\"resources\":{}}}" version="&Version{Version:v3.4.4,BuildDate:2022-11-29T16:49:53Z,GitCommit:3b2626ff900aff2424c086a51af5929fb0b2d7e5,GitTag:v3.4.4,GitTreeState:clean,GoVersion:go1.18.8,Compiler:gc,Platform:linux/amd64,}"
lovely-rhino-f9r5l-cowsay-100404247: time="2023-01-24T07:39:54.522Z" level=info msg="Starting deadline monitor"
lovely-rhino-f9r5l-cowsay-100404247: time="2023-01-24T07:39:56.523Z" level=info msg="Main container completed" error="<nil>"
lovely-rhino-f9r5l-cowsay-100404247: time="2023-01-24T07:39:56.523Z" level=info msg="No Script output reference in workflow. Capturing script output ignored"
lovely-rhino-f9r5l-cowsay-100404247: time="2023-01-24T07:39:56.523Z" level=info msg="No output parameters"
lovely-rhino-f9r5l-cowsay-100404247: time="2023-01-24T07:39:56.523Z" level=info msg="No output artifacts"
lovely-rhino-f9r5l-cowsay-100404247: time="2023-01-24T07:39:56.523Z" level=info msg="Alloc=7745 TotalAlloc=11920 Sys=18386 NumGC=3 Goroutines=7"
```

**Comment by sarabala1979** [comment]
@nikita-akuity This expression is evaluated in Expr library. Need to investigate how the string comparison on Expr.
Do you like to take a look?

**Comment by jiachengxu** [comment]
I will take a look at this issue

**Comment by jiachengxu** [comment]
I think the issue is about string comparison, I will investigate more, and for now a workaround could be removing the double quotes around `"Succeeded"`, so the ternary operator becomes something like this:
`'{{= workflow.status == Succeeded ? "SUCCESSFUL" : "FAILED" }}'`

**Comment by rdean150** [comment]
Would it be correct to assume that the fix implemented in that PR may also fix #8742 ? It would explain why exit lifecycle hooks ignore expressions based on ``workflow.status`` being any specific value, if ``workflow.status`` is always evaluating to ``nil`` in expr expressions when inside an exit handler.

**Comment by ksonbol** [comment]
I have the same issue in a dag with the ternry operator using
`value: '{{= tasks.some-task.status == "Succeeded" ? "success" : "error" }}'`
Is this also fixed in the mentioned commit?
